pkg/io/video/broadcast_test.go

```diff
@@ -0,0 +1,135 @@
+package video
+
+import (
+	"fmt"
+	"image"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func BenchmarkBroadcast(b *testing.B) {
+	var src Reader
+	img := image.NewRGBA(image.Rect(0, 0, 1920, 1080))
+	interval := time.NewTicker(time.Millisecond * 33) // 30 fps
+	defer interval.Stop()
+	src = ReaderFunc(func() (image.Image, error) {
+		<-interval.C
+		return img, nil
+	})
+
+	for n := 1; n <= 4096; n *= 16 {
+		n := n
+
+		b.Run(fmt.Sprintf("Readers-%d", n), func(b *testing.B) {
+			b.SetParallelism(n)
+			broadcaster := NewBroadcaster(src)
+			b.RunParallel(func(pb *testing.PB) {
+				reader := broadcaster.NewReader(false)
+				for pb.Next() {
+					reader.Read()
+				}
+			})
+		})
+	}
+}
+
+func TestBroadcast(t *testing.T) {
```

Comment:
I think it's better to add some more test conditions to simulate blocking in provider and consumer like:
https://github.com/pion/mediadevices/compare/add-broadcast-test-conditions-with-pause

Metrics of above branch:
```
=== RUN   TestBroadcast/SrcPause-false/DstPause-false/Readers-1
    TestBroadcast/SrcPause-false/DstPause-false/Readers-1: broadcast_test.go:164: Average FPS                      : 30.333333333333332
    TestBroadcast/SrcPause-false/DstPause-false/Readers-1: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-false/Readers-1: broadcast_test.go:166: Last frame count (src)           : 91
    TestBroadcast/SrcPause-false/DstPause-false/Readers-1: broadcast_test.go:167: Average last frame count (dst)   : 90
=== RUN   TestBroadcast/SrcPause-false/DstPause-false/Readers-16
    TestBroadcast/SrcPause-false/DstPause-false/Readers-16: broadcast_test.go:164: Average FPS                      : 30.333333333333325
    TestBroadcast/SrcPause-false/DstPause-false/Readers-16: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-false/Readers-16: broadcast_test.go:166: Last frame count (src)           : 91
    TestBroadcast/SrcPause-false/DstPause-false/Readers-16: broadcast_test.go:167: Average last frame count (dst)   : 90
=== RUN   TestBroadcast/SrcPause-false/DstPause-false/Readers-256
    TestBroadcast/SrcPause-false/DstPause-false/Readers-256: broadcast_test.go:164: Average FPS                      : 30.333333333333222
    TestBroadcast/SrcPause-false/DstPause-false/Readers-256: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-false/Readers-256: broadcast_test.go:166: Last frame count (src)           : 91
    TestBroadcast/SrcPause-false/DstPause-false/Readers-256: broadcast_test.go:167: Average last frame count (dst)   : 90
=== RUN   TestBroadcast/SrcPause-true/DstPause-false
=== RUN   TestBroadcast/SrcPause-true/DstPause-false/Readers-1
    TestBroadcast/SrcPause-true/DstPause-false/Readers-1: broadcast_test.go:164: Average FPS                      : 20.666666666666668
    TestBroadcast/SrcPause-true/DstPause-false/Readers-1: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-true/DstPause-false/Readers-1: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-true/DstPause-false/Readers-1: broadcast_test.go:167: Average last frame count (dst)   : 61
=== RUN   TestBroadcast/SrcPause-true/DstPause-false/Readers-16
    TestBroadcast/SrcPause-true/DstPause-false/Readers-16: broadcast_test.go:164: Average FPS                      : 20.666666666666668
    TestBroadcast/SrcPause-true/DstPause-false/Readers-16: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-true/DstPause-false/Readers-16: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-true/DstPause-false/Readers-16: broadcast_test.go:167: Average last frame count (dst)   : 61
=== RUN   TestBroadcast/SrcPause-true/DstPause-false/Readers-256
    TestBroadcast/SrcPause-true/DstPause-false/Readers-256: broadcast_test.go:164: Average FPS                      : 20.66666666666669
    TestBroadcast/SrcPause-true/DstPause-false/Readers-256: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-true/DstPause-false/Readers-256: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-true/DstPause-false/Readers-256: broadcast_test.go:167: Average last frame count (dst)   : 61
=== RUN   TestBroadcast/SrcPause-false/DstPause-true
=== RUN   TestBroadcast/SrcPause-false/DstPause-true/Readers-1
    TestBroadcast/SrcPause-false/DstPause-true/Readers-1: broadcast_test.go:164: Average FPS                      : 20.666666666666668
    TestBroadcast/SrcPause-false/DstPause-true/Readers-1: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-true/Readers-1: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-false/DstPause-true/Readers-1: broadcast_test.go:167: Average last frame count (dst)   : 61
=== RUN   TestBroadcast/SrcPause-false/DstPause-true/Readers-16
    TestBroadcast/SrcPause-false/DstPause-true/Readers-16: broadcast_test.go:164: Average FPS                      : 20.666666666666668
    TestBroadcast/SrcPause-false/DstPause-true/Readers-16: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-true/Readers-16: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-false/DstPause-true/Readers-16: broadcast_test.go:167: Average last frame count (dst)   : 61
=== RUN   TestBroadcast/SrcPause-false/DstPause-true/Readers-256
    TestBroadcast/SrcPause-false/DstPause-true/Readers-256: broadcast_test.go:164: Average FPS                      : 20.66666666666669
    TestBroadcast/SrcPause-false/DstPause-true/Readers-256: broadcast_test.go:165: Average dropped frames per second: 0
    TestBroadcast/SrcPause-false/DstPause-true/Readers-256: broadcast_test.go:166: Last frame count (src)           : 62
    TestBroadcast/SrcPause-false/DstPause-true/Readers-256: broadcast_test.go:167: Average last frame count (dst)   : 61
```

Reply:
I like it!